File: 01-Programming-Languages/C++/Algorithms/String/string_algorithms.cpp
/**
 * string_algorithms.cpp
 *
 * Common string algorithms implemented in C++ using std::string
 * and STL algorithms.
 *
 * Time complexities noted per function.
 */

#include <iostream>
#include <string>
#include <algorithm>
#include <cctype>
#include <unordered_map>
#include <sstream>

using namespace std;

/* -----------------------------------------------------------
 * Reverse a string (creates a new string).
 * Time:  O(n)
 * Space: O(n)
 * ----------------------------------------------------------- */
string reverse_string(const string &s) {
    string result(s.rbegin(), s.rend());
    return result;
}

/* -----------------------------------------------------------
 * In-place reverse.
 * Time:  O(n)
 * Space: O(1)
 * ----------------------------------------------------------- */
void reverse_in_place(string &s) {
    reverse(s.begin(), s.end());
}

/* -----------------------------------------------------------
 * Check whether a string is a palindrome.
 * Ignores case and non-alphanumeric characters.
 * Time:  O(n)
 * Space: O(1)
 * ----------------------------------------------------------- */
bool is_palindrome(const string &s) {
    int left = 0, right = (int)s.size() - 1;
    while (left < right) {
        while (left < right && !isalnum(s[left])) left++;
        while (left < right && !isalnum(s[right])) right--;
        if (tolower(s[left]) != tolower(s[right]))
            return false;
        left++;
        right--;
    }
    return true;
}

/* -----------------------------------------------------------
 * Count the number of vowels.
 * Time:  O(n)
 * Space: O(1)
 * ----------------------------------------------------------- */
int count_vowels(const string &s) {
    int count = 0;
    for (char c : s)
        if (string("aeiouAEIOU").find(c) != string::npos)
            count++;
    return count;
}

/* -----------------------------------------------------------
 * Longest common substring (dynamic programming).
 * Time:  O(m * n)
 * Space: O(m * n)
 * ----------------------------------------------------------- */
string longest_common_substring(const string &s1, const string &s2) {
    int m = s1.size(), n = s2.size();
    vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));
    int max_len = 0, end_idx = 0;

    for (int i = 1; i <= m; i++) {
        for (int j = 1; j <= n; j++) {
            if (s1[i - 1] == s2[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1] + 1;
                if (dp[i][j] > max_len) {
                    max_len = dp[i][j];
                    end_idx = i;
                }
            }
        }
    }
    return s1.substr(end_idx - max_len, max_len);
}

/* -----------------------------------------------------------
 * Run-length encoding compression.
 * "aaabbbcc" -> "a3b3c2", "abc" -> "abc" (no change).
 * Time:  O(n)
 * Space: O(n)
 * ----------------------------------------------------------- */
string compress_string(const string &s) {
    if (s.empty()) return "";

    ostringstream oss;
    int i = 0, n = s.size();

    while (i < n) {
        char current = s[i];
        int count = 0;
        while (i < n && s[i] == current) {
            count++;
            i++;
        }
        oss << current;
        if (count > 1)
            oss << count;
    }
    return oss.str();
}

/* -----------------------------------------------------------
 * First non-repeating character.
 * Time:  O(n)
 * Space: O(1) — alphabet size
 * ----------------------------------------------------------- */
int first_non_repeating(const string &s) {
    unordered_map<char, int> freq;
    for (char c : s)
        freq[c]++;
    for (int i = 0; i < (int)s.size(); i++)
        if (freq[s[i]] == 1)
            return i;
    return -1;
}

/* ============================================================
 * Demo
 * ============================================================ */
int main() {
    cout << "=== Reverse String ===" << endl;
    string s1 = "hello";
    cout << "Before: \"" << s1 << "\"" << endl;
    cout << "After:  \"" << reverse_string(s1) << "\"" << endl;
    reverse_in_place(s1);
    cout << "In-place: \"" << s1 << "\"" << endl;

    cout << "\n=== Palindrome Check ===" << endl;
    cout << "\"racecar\" -> " << boolalpha << is_palindrome("racecar") << endl;
    cout << "\"hello\"   -> " << is_palindrome("hello") << endl;
    cout << "\"A man, a plan, a canal: Panama\" -> "
         << is_palindrome("A man, a plan, a canal: Panama") << endl;

    cout << "\n=== Count Vowels ===" << endl;
    cout << "\"programming\" -> " << count_vowels("programming") << endl;
    cout << "\"AEIOU\"       -> " << count_vowels("AEIOU") << endl;

    cout << "\n=== Longest Common Substring ===" << endl;
    cout << "Between \"abcdef\" and \"zbcdf\"  -> \""
         << longest_common_substring("abcdef", "zbcdf") << "\"" << endl;
    cout << "Between \"GeeksforGeeks\" and \"GeeksQuiz\" -> \""
         << longest_common_substring("GeeksforGeeks", "GeeksQuiz") << "\"" << endl;

    cout << "\n=== String Compression ===" << endl;
    cout << "\"aaabbbcc\"  -> \"" << compress_string("aaabbbcc") << "\"" << endl;
    cout << "\"abc\"       -> \"" << compress_string("abc") << "\"" << endl;
    cout << "\"aabbbaa\"   -> \"" << compress_string("aabbbaa") << "\"" << endl;

    cout << "\n=== First Non-Repeating Character ===" << endl;
    string s2 = "leetcode";
    int idx = first_non_repeating(s2);
    cout << "\"leetcode\"  -> index " << idx << " ('" << (idx >= 0 ? s2[idx] : '?') << "')" << endl;

    return 0;
}

File: 01-Programming-Languages/Java/Algorithms/String/string_algorithms.java
package stringalgorithms;

/**
 * string_algorithms.java
 *
 * Common string algorithms implemented in Java using String and StringBuilder.
 *
 * Time complexities noted per method.
 */
public class string_algorithms {

    /* -----------------------------------------------------------
     * Reverse a string.
     * Time:  O(n)
     * Space: O(n)
     * ----------------------------------------------------------- */
    public static String reverseString(String s) {
        return new StringBuilder(s).reverse().toString();
    }

    /* -----------------------------------------------------------
     * Check whether a string is a palindrome.
     * Ignores case and non-alphanumeric characters.
     * Time:  O(n)
     * Space: O(1)
     * ----------------------------------------------------------- */
    public static boolean isPalindrome(String s) {
        int left = 0, right = s.length() - 1;
        while (left < right) {
            while (left < right && !Character.isAlphanumeric(s.charAt(left))) left++;
            while (left < right && !Character.isAlphanumeric(s.charAt(right))) right--;
            if (Character.toLowerCase(s.charAt(left)) != Character.toLowerCase(s.charAt(right)))
                return false;
            left++;
            right--;
        }
        return true;
    }

    /* -----------------------------------------------------------
     * Count the number of vowels in a string.
     * Time:  O(n)
     * Space: O(1)
     * ----------------------------------------------------------- */
    public static int countVowels(String s) {
        int count = 0;
        String vowels = "aeiouAEIOU";
        for (int i = 0; i < s.length(); i++) {
            if (vowels.indexOf(s.charAt(i)) >= 0)
                count++;
        }
        return count;
    }

    /* -----------------------------------------------------------
     * Longest common substring (dynamic programming).
     * Time:  O(m * n)
     * Space: O(m * n)
     * ----------------------------------------------------------- */
    public static String longestCommonSubstring(String s1, String s2) {
        int m = s1.length(), n = s2.length();
        int[][] dp = new int[m + 1][n + 1];
        int maxLen = 0, endIdx = 0;

        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (s1.charAt(i - 1) == s2.charAt(j - 1)) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                    if (dp[i][j] > maxLen) {
                        maxLen = dp[i][j];
                        endIdx = i;
                    }
                }
            }
        }
        return s1.substring(endIdx - maxLen, endIdx);
    }

    /* -----------------------------------------------------------
     * Run-length encoding compression.
     * "aaabbbcc" -> "a3b3c2", "abc" -> "abc"
     * Time:  O(n)
     * Space: O(n)
     * ----------------------------------------------------------- */
    public static String compressString(String s) {
        if (s == null || s.isEmpty()) return "";

        StringBuilder sb = new StringBuilder();
        int i = 0, len = s.length();

        while (i < len) {
            char current = s.charAt(i);
            int count = 0;
            while (i < len && s.charAt(i) == current) {
                count++;
                i++;
            }
            sb.append(current);
            if (count > 1) sb.append(count);
        }
        return sb.toString();
    }

    /* -----------------------------------------------------------
     * First non-repeating character index.
     * Time:  O(n)
     * Space: O(1) — fixed alphabet size
     * ----------------------------------------------------------- */
    public static int firstNonRepeating(String s) {
        int[] freq = new int[256];  /* ASCII frequency array */
        for (int i = 0; i < s.length(); i++)
            freq[s.charAt(i)]++;
        for (int i = 0; i < s.length(); i++)
            if (freq[s.charAt(i)] == 1)
                return i;
        return -1;
    }

    /* ============================================================
     * Demo
     * ============================================================ */
    public static void main(String[] args) {
        System.out.println("=== Reverse String ===");
        System.out.println("\"hello\" -> \"" + reverseString("hello") + "\"");

        System.out.println("\n=== Palindrome Check ===");
        System.out.println("\"racecar\" -> " + isPalindrome("racecar"));
        System.out.println("\"hello\"   -> " + isPalindrome("hello"));
        System.out.println("\"A man, a plan, a canal: Panama\" -> "
                + isPalindrome("A man, a plan, a canal: Panama"));

        System.out.println("\n=== Count Vowels ===");
        System.out.println("\"programming\" -> " + countVowels("programming"));
        System.out.println("\"AEIOU\"       -> " + countVowels("AEIOU"));

        System.out.println("\n=== Longest Common Substring ===");
        System.out.println("Between \"abcdef\" and \"zbcdf\"  -> \""
                + longestCommonSubstring("abcdef", "zbcdf") + "\"");
        System.out.println("Between \"GeeksforGeeks\" and \"GeeksQuiz\" -> \""
                + longestCommonSubstring("GeeksforGeeks", "GeeksQuiz") + "\"");

        System.out.println("\n=== String Compression ===");
        System.out.println("\"aaabbbcc\"  -> \"" + compressString("aaabbbcc") + "\"");
        System.out.println("\"abc\"       -> \"" + compressString("abc") + "\"");
        System.out.println("\"aabbbaa\"   -> \"" + compressString("aabbbaa") + "\"");

        System.out.println("\n=== First Non-Repeating Character ===");
        String s = "leetcode";
        int idx = firstNonRepeating(s);
        char ch = idx >= 0 ? s.charAt(idx) : '?';
        System.out.println("\"leetcode\" -> index " + idx + " ('" + ch + "')");
    }
}
